### ISSUE TITLE
Fix extension build on Mac

### ISSUE THREAD

**Issue by lassoan** [Slicer/SlicerJupyter#6]
Factory build fails on Mac with multiple errors, including

`Error copying file "libxeus.1.dylib" to "/.../SlicerJupyter-build/inner-build/_CPack_Packages/Darwin/TGZ/27269-macosx-amd64-SlicerJupyter-git2734bf9-2018-07-07/Slicer.app/Contents/lib/Slicer-4.9/libxeus.1.dylib".`

See full logs here: http://slicer.cdash.org/viewBuildError.php?buildid=1315187

**Comment by ihnorton** [comment]
Packaging succeeds if I manually change the libxeus reference in qSlicerJupyerKernel.dylib to an absolute path:

```
install_name_tool -change libxeus.1.dylib /opt/bld/SlicerJupyter-s5/lib/Slicer-4.9/libxeus.1.dylib /opt/bld/SlicerJupyter-s5/inner-build/lib/Slicer-4.9/qt-loadable-modules/libqSlicerJupyterKernelModule.dylib
```

**Comment by ihnorton** [comment]
The problem is xeus CMakeLists has [`set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)`](https://github.com/QuantStack/xeus/blob/dfca7176e3ae8cf880dab8786b71414238a8f215/CMakeLists.txt#L99). Adding `-DCMAKE_BUILD_WITH_INSTALL_RPATH:BOOL=FALSE` to the ExternalProject_Add is not enough, so we probably need to have upstream change it to only set that option conditionally.

**Comment by SylvainCorlay** [comment]
@ihnorton @jcfr we have made changes on the RPATH settings upstream.

It is not quite the same thing as what you were suggesting. Instead of changing `CMAKE_BUILD_WITH_INSTALL_RPATH` to `FALSE`, we set `MACOSX_RPATH` to `ON`.

**Comment by ihnorton** [comment]
@SylvainCorlay thanks. I confirmed that latest xeus master fixes the packaging problem for us, without the work-around in 3926d2431d366f0e0bff9480c7bff39be9f7b56c.

**Comment by SylvainCorlay** [comment]
Great!

The only thinf is that it does not fix the conda relocation issue. At least the cmake is getting simpler and cleaner.

**Comment by jcfr** [comment]
Thanks @SylvainCorlay for the help :smile:

**Comment by ihnorton** [comment]
> I confirmed that latest xeus master fixes the packaging problem for us, without the work-around in 3926d24.

Note: this comment was related to https://github.com/QuantStack/xeus/commit/74e5bd66d785866629519ccef5073095a08421fc --    which fixed our problem by hiding `CMAKE_MACOSX_INSTALL_RPATH` (because it was applied to a non-existent target)

The next xeus commit, https://github.com/QuantStack/xeus/commit/74e5bd66d785866629519ccef5073095a08421fc -- would break our packaging again because it now actually applies `MACOSX_RPATH` to the target.

SO, I suggest we keep the work-around for now in order that @SylvainCorlay can do what is needed to fix the conda packaging for xeus-cling.

**Comment by SylvainCorlay** [comment]
We apply MACOSX_RPATH but I think it is the right  behavior which will become the default in the future of cmake.

Did you try with the latest commit?

**Comment by ihnorton** [comment]
> Did you try with the latest commit?

Yes, I tried with https://github.com/QuantStack/xeus/commit/f26076bae01ae37d8840736c5f4ec6a6fef815df (I pasted the wrong one in the 2nd paragraph of my last comment)

> We apply MACOSX_RPATH but I think it is the right behavior which will become the default in the future of cmake.

I agree, but our current CMake packaging fix-up scripts don't handle it correctly right now (I've rewritten the fixup scripts for Slicer in Python, but we won't switch until after the current release).

**Comment by SylvainCorlay** [comment]
> I agree, but our current CMake packaging fix-up scripts don't handle it correctly right now (I've rewritten the fixup scripts for Slicer in Python, but we won't switch until after the current release).
>

Ok thanks. I just wanted to be sure that I am not doing something wrong here.

Unfortunately this does not resolve the conda issue. I feel that solving it will require us to figure out internals of conda-build, cmake and how osx handles rpaths...

**Comment by jcfr** [comment]
To provide some background, within Slicer we expect all rpath to be full paths so that our "fixup" script can traverse the full dependency tree and know which library belong to where. 

With modern CMake and the recent work of @ihnorton , we should be able to streamline the process.
